File: Python/binary-tree-maximum-path-sum.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
INT_MIN = -2147483648
class Solution:
    # @param root, a tree node
    # @return an integer
    def __init__(self):
        import sys
        self.max_path_sum = -sys.maxsize-1
        
    def maxPathSum(self, root):
        self.get_max(root)
        return self.max_path_sum
        
    def get_max(self, root):
        if not root:
            return 0
        left = self.get_max(root.left)
        right = self.get_max(root.right)
        
        sum_all = root.val
        
        if left>0:
            sum_all += left
        if right>0:
            sum_all += right
            
        self.max_path_sum = max(self.max_path_sum, sum_all)
        
        return max(0, left, right)+root.val

File: C++/binary-tree-maximum-path-sum.cpp
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
private:
    int max_path_sum = INT_MIN;
    int path_sum(TreeNode* root) {
        if(!root) return 0;
        int left = path_sum(root->left);
        int right = path_sum(root->right);
        int sum = root->val;
        if(left>0) sum += left;
        if(right>0) sum += right;
        max_path_sum = std::max(max_path_sum, sum);
        return std::max(0, std::max(left, right)) + root->val;
    }
public:
    int maxPathSum(TreeNode* root) {
        path_sum(root);
        return max_path_sum;
    }
};

